README.md
# Census Management System
A system for managing and searching the data released by Word Bank.


Data/Changes
Kenya 2000 male 1000
Ghana 2010 male 1


src/Pair.java
import java.io.Serializable;

public class Pair<K extends Comparable<K>,T extends Comparable<T>> implements Comparable<Pair<K,T>>,Serializable
{
	static final long serialVersionUID = 42L;
	@Override
	public boolean equals(Object o)
	{
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;

		Pair pair = (Pair) o;

		if (first != null ? !first.equals(pair.first) : pair.first != null) return false;
		if (second != null ? !second.equals(pair.second) : pair.second != null) return false;

		return true;
	}

	@Override
	public int hashCode()
	{
		int result = first != null ? first.hashCode() : 0;
		result = 31 * result + (second != null ? second.hashCode() : 0);
		return result;
	}

	public K first;
	public T second;
	public Pair(K first, T second)
	{
		this.first = first;
		this.second = second;
	}

	@Override
	public int compareTo(Pair<K,T> o)
	{
		if (first.equals(o.first)==false)
			return first.compareTo(o.first);
		return second.compareTo(o.second);
	}
}


src/Sheet.java
/**
 * Created with IntelliJ IDEA.
 * User: LGM
 * Date: 2/24/16
 * Time: 3:00 PM
 */
public class Sheet
{
	int rows,cols;
	String [][] values;

	public Sheet(int rows, int cols, String[][] values)
	{
		this.rows = rows;
		this.cols = cols;
		this.values = values;
	}

	public int getCols()
	{
		return cols;
	}

	public void setCols(int cols)
	{
		this.cols = cols;
	}

	public int getRows()
	{
		return rows;
	}

	public void setRows(int rows)
	{
		this.rows = rows;
	}

	public String[][] getValues()
	{
		return values;
	}

	public void setValues(String[][] values)
	{
		this.values = values;
	}

	public Pair<Integer, Integer> getCell(String country, int year)
	{
		for(int i=0; i<getCols(); i++)
			if(values[16][i].equals(Integer.toString(year)))
			{
				for(int j=0; j<getRows(); j++)
					if(values[j][2].equals(country))
						return new Pair<>(j,i);
			}
		return null;
	}
}


src/Lister.java
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;

/**
 * Created by qizilbash on 3/1/16.
 */
public class Lister {

    private int year;
    private int countary = 2;

    private Double[] arr;
    private ArrayList<String> list = new ArrayList<String>();
    private Sheet sheet;

    private ImportData importData = new ImportData();

    private String str1 ="Data/WPP2015_POP_F01_2_TOTAL_POPULATION_MALE.XLS";
    private String str2 ="Data/WPP2015_POP_F01_3_TOTAL_POPULATION_FEMALE.XLS";
    private String str3 ="Data/WPP2015_POP_F02_POPULATION_GROWTH_RATE.XLS";
    private String str4 ="Data/WPP2015_POP_F03_RATE_OF_NATURAL_INCREASE.XLS";

    public  ArrayList<String> getList(int year, int opt) throws IOException {
        this.year = year;
        this.year -=1950;
        this.year +=5;
        switch (opt){
            case 1 :
                sheet = importData.Import(str1,0);
                break;
            case 2 :
                sheet = importData.Import(str2,0);
                break;
            case 3 :
                sheet = importData.Import(str3,0);
                break;
            case 4 :
                sheet = importData.Import(str4,0);
                break;

        }
        arr = new Double[sheet.getRows()-19];
        System.out.println(sheet.getRows());
        for (int i=0;i<sheet.getRows()-19;i++){
            arr[i]= Double.valueOf(sheet.getValues()[i+19][this.year]);
        }

        Arrays.sort(arr);
        System.out.println(arr.length);
        for (double item : arr){
            for (int i =0 ; i< sheet.getRows()-19;i++){
                if (Double.valueOf(sheet.getValues()[i+19][this.year]) == item){
                    list.add(sheet.getValues()[i][countary]);
                    break;
                }
            }
        }

        for (String item : list)
            System.out.println(item);

        return list;
    }


    public static void main(String[] args) throws IOException {
        new Lister().getList(1950,1);
    }


}


src/PDFWithText.java
/**
 * @(#)PDFWithText.java
 *
 *
 * @author Stephen H
 * @version 1.00 2011/6/21
 */

import org.apache.pdfbox.pdmodel.PDDocument;
import org.apache.pdfbox.pdmodel.PDPage;
import org.apache.pdfbox.pdmodel.*;
import org.apache.pdfbox.pdmodel.font.PDType1Font;
import org.apache.pdfbox.pdmodel.font.PDFont;

public class PDFWithText {
    public static void main(String[] args) {
        PDDocument doc = null;
        PDPage page = null;

       try{
           doc = new PDDocument();
           page = new PDPage();

           doc.addPage(page);
           PDFont font = PDType1Font.HELVETICA_BOLD;

           PDPageContentStream content = new PDPageContentStream(doc, page);
           content.beginText();
           content.setFont( font, 12 );
           content.moveTextPositionByAmount( 100, 700 );
           content.drawString("Csalam khubi che khabar felans");

           content.endText();
           content.close();
          doc.save("PDFWithText.pdf");
          doc.close();
    } catch (Exception e){
        System.out.println(e);
    }
}
}

src/ImportData.java
import com.sun.org.apache.xpath.internal.SourceTree;
import org.apache.poi.hssf.usermodel.HSSFCell;
import org.apache.poi.hssf.usermodel.HSSFRow;
import org.apache.poi.hssf.usermodel.HSSFSheet;
import org.apache.poi.hssf.usermodel.HSSFWorkbook;
import org.apache.poi.poifs.filesystem.POIFSFileSystem;

import java.io.*;
import java.util.HashSet;
import java.util.Scanner;
import java.util.Set;

/**
 * Created with IntelliJ IDEA.
 * User: LGM
 * Date: 2/24/16
 * Time: 2:38 PM
 */

public class ImportData
{
	public String GetData(Sheet s, String country, int year)
	{
		for(int i=0; i<s.getCols(); i++)
			if(s.values[16][i].equals(Integer.toString(year)))
			{
				for(int j=0; j<s.getRows(); j++)
					if(s.values[j][2].equals(country))
						return s.values[j][i];
			}
		return null;
	}
	public Sheet Import(String File, int sheetNum) throws IOException
	{
		POIFSFileSystem fs = new POIFSFileSystem(new FileInputStream(File));
		HSSFWorkbook wb = new HSSFWorkbook(fs);
		if(sheetNum>wb.getNumberOfSheets()) return null;
		HSSFSheet sheet = wb.getSheetAt(sheetNum);
		HSSFRow row;
		HSSFCell cell;

		int rows; // No of rows
		rows = sheet.getPhysicalNumberOfRows();
		int cols = 0; // No of columns
		int tmp = 0;

		// This trick ensures that we get the data properly even if it doesn't start from first few rows
		for(int i = 0; i < 10 && i < rows; i++) {
			row = sheet.getRow(i);
			if(row != null) {
				tmp = sheet.getRow(i).getPhysicalNumberOfCells();
				if(tmp > cols) cols = tmp;
			}
		}
		String [][] res=new String[rows][cols];
		for(int r = 0; r < rows; r++) {
			res[r]=new String[cols];
			row = sheet.getRow(r);
			if(row != null) {
				for(int c = 0; c < cols; c++) {
					cell = row.getCell((short)c);
					res[r][c]=(cell==null?null:cell.toString());
				}
			}
		}
		return new Sheet(rows,cols,res);
	}
	static Sheet male,female;
	static Set<String> protect;
	public static void main(String[] args) throws IOException
	{
		protect=new HashSet<>();
		ImportData importData=new ImportData();
		male=importData.Import("Data/WPP2015_POP_F01_2_TOTAL_POPULATION_MALE.XLS",0);
		female=importData.Import("Data/WPP2015_POP_F01_3_TOTAL_POPULATION_FEMALE.XLS",0);
		File f = new File("Data/Changes");
		if (f.exists())
		{
			Scanner changes = new Scanner(f);
			while (changes.hasNext())
			{
				String country = changes.next();
				int year = changes.nextInt();
				String s = changes.next();
				Pair<Integer, Integer> cell = (s.equals("male") ? male : female).getCell(country, year);
				if (cell != null)
					(s.equals("male") ? male : female).values[cell.first][cell.second] = changes.next();
				System.err.println("updated");
			}
			changes.close();
		}
		File f2 = new File("Data/Protect");
		if (f2.exists())
		{
			Scanner changes = new Scanner(f2);
			while (changes.hasNext())
			{
				String country = changes.next();
				protect.add(country);
			}
			changes.close();
		}
		PrintStream change=new PrintStream(new FileOutputStream("Data/Changes",true));
		PrintStream protectLog=new PrintStream(new FileOutputStream("Data/Protect",true));

		Scanner scanner=new Scanner(System.in);
		while (true)
		{
			String com = scanner.next();
			if (com.equals("1"))
			{
				String country=scanner.next();
				int year=scanner.nextInt();
				System.out.println(importData.GetData(male,country,year)+" , "+importData.GetData(female,country,year));
			}
			if (com.equals("2"))
			{
				String country=scanner.next();
				int year=scanner.nextInt();
				String s=scanner.next();
				Pair<Integer,Integer> cell=(s.equals("male")?male:female).getCell(country,year);
				String value=scanner.next();
				if (protect.contains(country))
					continue;
				if (cell!=null)
					(s.equals("male")?male:female).values[cell.first][cell.second]=value;
				change.println(country+" "+year+" "+s+" "+value);
			}
			if (com.equals("3"))
			{
				int year=scanner.nextInt();
				int sheetNo=scanner.nextInt();
				new Lister().getList(year,sheetNo);
			}
			if (com.equals("4"))
						{
							int year=scanner.nextInt();
							int sheetNo=scanner.nextInt();
							new Lister().getList(year,sheetNo);
						}
			if (com.equals("8"))
			{
				String country=scanner.next();
				if (!protect.contains(country))
				{
					protect.add(country);
					protectLog.println(country);
				}
			}
		}
	}
}
